src/pip/_internal/index/sources.py

```diff
@@ -71,6 +72,8 @@ def _scan_directory(self) -> None:
                 try:
                     project_filename = parse_sdist_filename(entry.name)[0]
                 except InvalidSdistFilename:
+                    if entry.name.endswith("-variants.json"):
+                        self.variants_json = entry.path
```

Comment:
Doesn't this class have to account for multiple versions? If yes, then you'd probably need a dict from package names+versions to paths, similarly to how we do it in the regular codepath.